sql/core/src/test/scala/org/apache/spark/sql/streaming/StreamingQueryOptimizationCorrectnessSuite.scala

```diff
@@ -416,4 +416,39 @@ class StreamingQueryOptimizationCorrectnessSuite extends StreamTest {
       )
     }
   }
+
+  test("SPARK-48267: regression test, stream-stream union followed by stream-batch join") {
+    withTempDir { dir =>
+      val input1 = MemoryStream[Int]
+      val input2 = MemoryStream[Int]
+
+      val df1 = input1.toDF().withColumn("code", lit(1))
```

Comment:
Given the sequence of optimizations in reproducer, it wouldn't happen if the union wasn't eliminated, as the value of 'code' won't be known during optimization phase. That said, if we call observe() or withWatermark() or etc in these streaming DF, the bug wouldn't be triggered.